### PR TITLE
feat: add dtekv-upload and dtekv-download tools to navigation

### PR DESCRIPTION
**Pull request by Alvinn8**
Awesome emulator btw!

This PR adds buttons to be able to upload and download data to/from memory. It's useful for projects like ours that require uploading certain files/data to the board before running the binary.

### CODE REVIEW

File: src/cpu.ts [hunk 1]
```diff
@@ -112,7 +112,7 @@ function cpuLoop() {
 
 export function loadBinary(binary: Uint8Array) {
   currentLoadedBinary = new Uint8Array(binary);
-  cpu.set_to_new();
+  // cpu.set_to_new();
```
Comment (Alvinn8) [inline src/cpu.ts:115]: While reusing the CPU instance might seem a bit weird, our project sort of "relies" on leaking memory between runs. Our code assumes that files have already been uploaded when the program starts running, so therefore we need the memory to be populated before the binary is ran. The user can always reload the page to get a full reset. I'm open to other solutions though if you know a better way to make this work.

**Comment by Alvinn8** [comment]
@PumpedSardines ready for review :)

**Comment by PumpedSardines** [comment]
So awesome seeing PR:s to this project 🤩

I'm really busy today, but I'll look at this more in depth tomorrow. 

My first thought is having two buttons for reset, one for hard reset and one for the more softer reset provided in this PR. Not sure how to have both buttons though without causing confusion to other users... would love to hear your input

File: src/pages/Emulator/views/Nav/index.tsx [hunk 2]
```diff
@@ -144,4 +154,175 @@ function ExampleButton() {
...
+                alert(e);
+                return;
+              }
+              const blob = new Blob([data]);
+              downloadBlob(blob, fileName);
+            }}
+          >
+            Download
+          </button>
+        </div>
```
Comment (PumpedSardines) [inline src/pages/Emulator/views/Nav/index.tsx:311]: Same validation here

File: wasm/src/lib.rs [hunk 3]
```diff
@@ -170,4 +170,20 @@ impl Cpu {
...
+        let mut buf: Vec<u8> = Vec::with_capacity(length);
+        let sdram = self.sdram.borrow();
+        for i in 0..length {
+            let byte = sdram
+                .load_byte(addr + i as u32)
+                .map_err(|_| JsError::new("Out of range."))?;
+            buf.push(byte);
+        }
+        Ok(buf)
+    }
```
Comment (PumpedSardines) [inline wasm/src/lib.rs:188]: Note to self, we should add a read_data_at in the core library. Outside the scope of this PR

File: src/pages/Emulator/views/Nav/index.tsx [hunk 4]
```diff
@@ -89,27 +91,35 @@ function Nav() {
   );
 }
 
-function ExampleButton() {
-  const [exampleButtonActive, setExampleButtonActive] = useState(false);
-  const exampleRef = useRef<HTMLDivElement>(null);
-  const exampleButtonRef = useRef<HTMLButtonElement>(null);
-
+function useCloseOnClickOutside(
```
Comment (PumpedSardines) [inline src/pages/Emulator/views/Nav/index.tsx:94]: Maybe make this more generic and move this into the `hooks/` directory

**Comment by Alvinn8** [comment]
I added some input validation. I haven't used `setCustomValidity` before, so see if I did it somewhat correctly 😅.

I made it perform hard resets when loading examples so that the screen resets etc. ButI kept the soft reset for loading binaries for now.

Having two buttons sounds like a good idea, communicating which button does what to the users might be harder. Maybe hovering the "Reset" button can reveal a button underneath, "Hard reset", "Full reset", or "Reset memory" maybe? Or if the hard reset should be the default and the soft reset a secondary button. I'm not sure what's best.

**Review by PumpedSardines — APPROVED** [review]

**Pull request merged**